cloudfour/pwa-workshop
# Feature: Service worker exercise 4

## Overview

This PR introduces the service worker logic for exercise 4. Additions include:
- add a `cacheFirst()` caching strategy function
    - fetch from cache
        - return a cached response if found
    - fallback to fetch from the network
        - store a copy of the network response in the cache
        - return the network response
- add an offline page
- add an image fallback
- new images
- add page 1, 2, 3 HTML files

## Testing

1. Open a Chrome incognito browser window

1. Open the DevTools and check the **Disable cache** checkbox in the **Network** tab. This will allow us to bypass the browser cache and see either network or SW cache responses

1. Load up https://deploy-preview-18--festive-keller-9a29cb.netlify.com/page-1.html, the first page load should be all from the network:
    <img width="1552" alt="Screen Shot 2019-06-01 at 2 19 54 PM" src="https://user-images.githubusercontent.com/459757/58753697-6364c080-8478-11e9-984a-b93eb0fbd405.png">

1. Open the **Application** > **Cache Storage** panel and confirm the `/offline.html` and `/images/fallback.svg` assets were precached

    <img width="862" alt="Screen Shot 2019-06-01 at 2 21 48 PM" src="https://user-images.githubusercontent.com/459757/58753741-dff79f00-8478-11e9-917c-5df803b56d56.png">


1. Navigate to **Page 2** and confirm the following assets were added to the **pwa-workshop-v1** cache: 
  
    - `/page-2.html`
    - `/scripts/register-service-worker.js`

        <img width="860" alt="Screen Shot 2019-06-01 at 2 26 52 PM" src="https://user-images.githubusercontent.com/459757/58753795-88a5fe80-8479-11e9-8402-de740ea92fc4.png">

1. Navigate back to **Page 1** and confirm the following assets were added to the **pwa-workshop-v1** cache: 
    - `/page-1.html`
    - `/images/gerardo-r2-150x150-c-default.png`
    - `/images/homework.svg`
    - `/images/jason-150x150-c-default.png`

        <img width="860" alt="Screen Shot 2019-06-01 at 2 32 05 PM" src="https://user-images.githubusercontent.com/459757/58753838-521cb380-847a-11e9-93d1-c3e6632714f3.png">

1. In DevTools **Network** tab, check the **Offline** checkbox to go offline

    <img width="775" alt="Screen Shot 2019-06-01 at 2 39 44 PM" src="https://user-images.githubusercontent.com/459757/58753898-3665dd00-847b-11e9-9e0a-cb4a18be718d.png">


1. Navigate to **Page 2** and confirm the page loads with all its assets while offline

    ![localhost_8000_page-2 html](https://user-images.githubusercontent.com/459757/58753913-8a70c180-847b-11e9-91c3-9c00ec39f524.png)

1. Navigate back to **Page 1** and confirm the page loads with all its assets while offline

    ![localhost_8000_page-1 html](https://user-images.githubusercontent.com/459757/58753919-b2f8bb80-847b-11e9-8bfd-fb48f39b8d69.png)

1. Navigate to **Page 3** and confirm the **Offline** page loads (Page 3 hasn't been cached yet)

    ![localhost_8000_page-3 html](https://user-images.githubusercontent.com/459757/58753925-d4f23e00-847b-11e9-8149-0c97d4d5eee5.png)

1. Navigate between **Page 1**, **Page 2** and **Page 3** and confirm each continues to load while offline (**Page 3** should be the **Offline** page)

1. From the **Application** > **Cache Storage** > **pwa-workshop-v1** cache, delete the `/images/homework.svg` file

    <img width="817" alt="Screen Shot 2019-06-01 at 2 47 43 PM" src="https://user-images.githubusercontent.com/459757/58753968-9a3cd580-847c-11e9-8977-0719f2f59144.png">

1. Navigate to **Page 1** and confirm the fallback `/images/fallback.svg` image is loaded in place of the no longer cached `/images/homework.svg` image

    ![localhost_8000_page-1 html (1)](https://user-images.githubusercontent.com/459757/58753982-e8ea6f80-847c-11e9-8318-bd609970d7c5.png)

1. From the **Network** tab, allow your device to go back online by _unchecking_ the **Offline** checkbox

    ![go-back-online](https://user-images.githubusercontent.com/459757/58754010-2c44de00-847d-11e9-89a0-a42b4c8a9cdb.gif)

1. Refresh **Page 1** and confirm the `/images/homework.svg` image loads and that it is cached again in the **pwa-workshop-v1** cache

    <img width="688" alt="Screen Shot 2019-06-01 at 2 55 38 PM" src="https://user-images.githubusercontent.com/459757/58754020-58605f00-847d-11e9-9eff-aa463a7f7bb9.png">

    <img width="883" alt="Screen Shot 2019-06-01 at 2 56 30 PM" src="https://user-images.githubusercontent.com/459757/58754036-8a71c100-847d-11e9-831d-e40eafd0a998.png">

1. While **online**, navigate to **Page 3** and confirm the **Offline** page no longer loads

    ![localhost_8000_page-3 html (1)](https://user-images.githubusercontent.com/459757/58754047-bab95f80-847d-11e9-847f-c4d74684302e.png)


1. Confirm the `/page-3.html` file was added to the **pwa-workshop-v1** cache:

    <img width="897" alt="Screen Shot 2019-06-01 at 3 00 01 PM" src="https://user-images.githubusercontent.com/459757/58754064-fc4a0a80-847d-11e9-8f08-1c2d94efe723.png">

1. Go offline again via the **Network** > **Offline** checkbox

    ![go-offline](https://user-images.githubusercontent.com/459757/58754088-5d71de00-847e-11e9-9d97-258eb9bc366c.gif)

1. Navigate between **Page 1**, **Page 2** and **Page 3** and confirm each page loads all of its assets (no **Offline** page should load this time because all three pages are cached)

---

- [Trello Card](https://trello.com/c/uMFcUyaS/46-service-worker-exercise-4-code)
